### ISSUE TITLE
Support local audio and video files

### ISSUE THREAD

**Issue by shun-liang** [shun-liang/yt2doc#29]
https://www.reddit.com/r/youtubedl/comments/1g574wr/comment/lsc7j6y/?context=3

> It would be great if you could also point it at a local directory and transcribe into this nice format. I've been slowly working on something similar, I've got a directory of videos and I've downloaded whisper and started building a docker container for it to process.

https://www.reddit.com/r/DataHoarder/comments/1g4342q/comment/lsc8v1h/?context=3

> Can I ask if there is any interest or plans to implement your program with local folders of stuff?
> 
> https://postimg.cc/xqg6Y11D
> 
> For additional context, I have folders of video lecture series. The videos do have subtitles but they are low-effort auto-generated and are wrong regularly on tech-related words. Whisper, on the other hand, works far better.

**Comment by shun-liang** [comment]
Loading the local video and audio files and just pipe them to Whisper and all the down streaming processing would be trivial.

However, to generate more accurate readable Whisper transcriptions (e.g. with punctuations), prompts with the context of the videos and audios need to be provided to Whisper through [prompts](https://cookbook.openai.com/examples/whisper_prompting_guide). For online video and audio sources (e.g. YouTube, Apple Podcasts) yt2doc [sanitizes](https://github.com/shun-liang/yt2doc/blob/5f186afe3ca567dc11d0d5aaf881cda000bcf706/src/yt2doc/transcription/transcriber.py#L74-L85) the title and the description and feed them into Whisper as prompts. I don't think local files have that amount of context available.

I think there is a possible solution which is to send the file names of the local videos or audios to a local LLM and asks it to generate a few fictitious sentences, as the OpenAI Whisper coobook [suggests](https://cookbook.openai.com/examples/whisper_prompting_guide#fictitious-prompts-can-be-generated-by-gpt). Will experiment with that idea.

**Comment by xorrosive** [comment]
> Loading the local video and audio files and just pipe them to Whisper and all the down streaming processing would be trivial.
> 
> However, to generate more accurate readable Whisper transcriptions (e.g. with punctuations), prompts with the context of the videos and audios need to be provided to Whisper through [prompts](https://cookbook.openai.com/examples/whisper_prompting_guide). For online video and audio sources (e.g. YouTube, Apple Podcasts) yt2doc [sanitizes](https://github.com/shun-liang/yt2doc/blob/5f186afe3ca567dc11d0d5aaf881cda000bcf706/src/yt2doc/transcription/transcriber.py#L74-L85) the title and the description and feed them into Whisper as prompts. I don't think local files have that amount of context available.
> 
> I think there is a possible solution which is to send the file names of the local videos or audios to a local LLM and asks it to generate a few fictitious sentences, as the OpenAI Whisper coobook [suggests](https://cookbook.openai.com/examples/whisper_prompting_guide#fictitious-prompts-can-be-generated-by-gpt). Will experiment with that idea.

Hello, an idea to experiment with for providing context for offline videos could be description pages belonging to the video(s). For example, in the case of courses, there is usually a sales page or a course description or at least a "syllabus" attached (think, Udemy).

Otherwise getting context from a descriptive filename, and possibly asking the user to describe in just a few words what the video is about, could potentially prove helpful. I'm not completely sure, but just throwing it out there.

**Comment by shun-liang** [comment]
> description pages belonging to the video(s).

Potentially. Though I am not sure yet about what the interface should look like, to support this but without massively bloating the scope of this project. Potentially we can open a command line option or a parameter `whisper_prompt` (as yt2doc can technically be used as a Python library), and the logic of crawling and parsing the web page into whisper prompt can be done somewhere else as a seperate project.

**Comment by xorrosive** [comment]
> > description pages belonging to the video(s).
> 
> Potentially. Though I am not sure yet about what the interface should look like, to support this but without massively bloating the scope of this project. Potentially we can open a command line option or a parameter `whisper_prompt` (as yt2doc can technically be used as a Python library), and the logic of crawling and parsing the web page into whisper prompt can be done somewhere else as a seperate project.

A "whisper prompt" passed as an argument sounds great. Max results for least effort.

**Comment by mrfragger** [comment]
> However, to generate more accurate readable Whisper transcriptions (e.g. with punctuations), prompts with the context of the videos and audios need to be provided to Whisper through [prompts](https://cookbook.openai.com/examples/whisper_prompting_guide). 

yeah prompt makes a huge difference.  In my experience I started out generating audiobooks and found the limitation of a wave file. is 34 hours or was it 36 hours ..can't remember.  Anyway there would be hallucinations of 20 mins sometimes even 30 minutes of just repeating phrases.  So I wrote substitcher which was to split audiobook into 999 audio segments and whisper.cpp to transcribe each one then just cat all the vtt subs back together into one sub.  Subs would get outta synch around the 200 mark or so.  So limited to 150 then eventually 200.  About 2 months ago bumped it up to 400 cuz I provide a 5ms or 6ms offset to each sub.  Now I got it up to 2,000 segments and use a 6.5ms offset and it works great.  Usually divide segments into 2min or 3mins.  This way even if it goes bonkers with repeating phrases usually it's only less than a minute of damage done.  Large-v2 (get 2x realtime) does a better job of avoiding them but I like large-turbo-v3 as gets 5x realtime speed.  Puncuation works great and post-processing I always run it to verify any overlapping subs.

for f in *.vtt
do ffmpeg -y -i "$f" "${f%.*}"_overlapping.srt
done
for f in *_overlapping.srt
do awk '
  BEGIN {
    RS = "";
    OFS = FS = "\n";
    getline;
    n = split($0, prev_rec);
    split($2, prev_time, / --> /);
  }
  {
    split($2, a, / --> /);
    if (a[1] < prev_time[2])
      prev_rec[2] = prev_time[1]" --> "a[1];
    for (i=1;i<=n;i++)
      print prev_rec[i];
    printf("\n");
    n = split($0, prev_rec);
    split($2, prev_time, / --> /)
  }
  END {
    print
  }' "$f" > "${f%_overlapping.*}"_overlapfixed.srt
done
for f in *_overlapfixed.srt
do ffmpeg -y -i "$f" "${f%.*}".vtt
done 
for f in *_overlapping.srt
do git diff --no-index --word-diff=color --unified=5 --word-diff-regex=. "$f" "${f%_overlapping.*}"_overlapfixed.srt | aha --black -w -y 'font-size:1.3em' -t "Fixed Overlapping Timecodes" > "${f%_overlapping.*}"_diff.html
done
rm -f *_overlapping.srt
rm -f *_overlapfixed.srt
rm *.srt
cd ..

then also do a proper noun in case it missed capitalizing any cities, countries, first names, last names, etc.

then check for repeating words with some exceptions of course
    "Remove repeating words - happy happy, care care care, no no no no")
    echo "Remove repeating words for vtt subtitles in substitcher directory"
    echo ""
    read -p "Press ENTER to remove repeating words or m (Main Menu) " choice
    [[ "$choice" == [Mm]* ]] && exec "$0" || echo "converting ..."
    dt=$( date +%Y_%m_%d_%H_%M_%S)
    [ ! -d output ] && mkdir output
    mkdir output/"$dt"
    cp *.vtt output/"$dt"
    cd output/"$dt"
    for f in *.vtt
do gsed -E \
-e "s|\bno no\b|no333no|g" \
-e "s|\bha ha\b|ha333ha|g" \
-e "s|\bher her\b|her333her|g" \
-e "s|\bbye bye\b|byebye|g" \
-e "s|\breally really\b|reallyreally|g" \
-e "s|\bstuff stuff\b|stuffstuff|g" \
-e "s|\bmany many\b|manymany|g" \
-e "s|\bvery very\b|veryvery|g" \
-e "s|\bhad had\b|hadhad|g" \
-e "s|\bblah blah\b|blahblah|g" \
-e "s|\bilai ilai\b|ilaiilai|g" \
-e "s|\bre re\b|rere|g" \
-e "s|\bthis this\b|thisthis|g" \
-e "s|\bthat that\b|thatthat|g" \
-e "s|\bever ever\b|everever|g" \
-e "s|\byou you\b|youyou|g" \
-e "s|\bso so\b|soso|g" \
-e "s|\bbeing being\b|beingbeing|g" \
-e "s|\b(\w+)\ \1\b|\1|g" \
-e "s|\b(\w+)\ \1\b|\1|g" \
-e "s|\b(\w+)\ \1\b|\1|g" \
-e "s|no333no|no no|g" \
-e "s|ha333ha|ha ha|g" \
-e "s|her333her|her her|g" \
-e "s|byebye|bye bye|g" \
-e "s|reallyreally|really really|g" \
-e "s|stuffstuff|stuff stuff|g" \
-e "s|manymany|many many|g" \
-e "s|veryvery|very very|g" \
-e "s|hadhad|had had|g" \
-e "s|blahblah|blah blah|g" \
-e "s|ilaiilai|ilai ilai|g" \
-e "s|rere|re re|g" \
-e "s|thisthis|this this|g" \
-e "s|thatthat|that that|g" \
-e "s|everever|ever ever|g" \
-e "s|youyou|you you|g" \
-e "s|soso|so so|g" \
-e "s|beingbeing|being being|g" \
 "$f" > "repeats_$f" ; \
git diff --no-index --word-diff=color --unified=2 --word-diff-regex=. "$f" "repeats_$f" \
| aha --black -t 'Repeating Words' -w -y 'font-size:1.3em' > "repeats_${f%.*}.html" ; \
rm "$f" ;  done
    cd ../..
